sycl/source/detail/sampler_impl.cpp

```diff
@@ -95,10 +94,14 @@ sampler_impl::~sampler_impl() {
 }
 
 ur_sampler_handle_t
-sampler_impl::getOrCreateSampler(const ContextImplPtr &ContextImpl) {
+sampler_impl::getOrCreateSampler(context_impl &ContextImpl) {
+  // Just for the `MContextToSampler` lookups. Could probably be changed once we
+  // move to C++20 and would have heterogeneous lookup.
+  std::shared_ptr<context_impl> ContextImplPtr = ContextImpl.shared_from_this();
```

Comment:
This is extra overhead, but we have unconditional mutex lock below (and another one near return on non-exceptional code path), so shouldn't matter.